### PR TITLE
TVB-2587: Support upoaders in rest launch operation

### PR DESCRIPTION
**Pull request by VinczeRobert**
I closed my previous pull request https://github.com/the-virtual-brain/tvb-root/pull/40 and I made a new one, this time from a fork.

I also added an example for launch_operation.

I am yet to make the changes that we talked about with Lia today, 13 February.

### CODE REVIEW

**Comment by liadomide** [comment]
I made 2 suggestions on PR#40 which are still valid here: about `utils.py` and the `page_size` constant. Pls consider also

File: framework_tvb/tvb/interfaces/rest/client/operation/operation_api.py [hunk 1]
```diff
@@ -54,16 +54,26 @@ def get_operations_results(self, operation_gid):
         return response, DataTypeDto
 
     @handle_response
-    def launch_operation(self, project_gid, algorithm_module, algorithm_classname, view_model, temp_folder):
+    def launch_operation(self, project_gid, algorithm_module, algorithm_classname, view_model, temp_folder, data_file_path_1=None, data_file_path_2=None):
```
Comment (liadomide) [inline framework_tvb/tvb/interfaces/rest/client/operation/operation_api.py]: Why don't we already have a list/dict of files here?

File: framework_tvb/tvb/interfaces/rest/client/tvb_client.py [hunk 2]
```diff
@@ -106,13 +107,15 @@ def fire_simulation(self, project_gid, session_stored_simulator):
         """
         return self.simulation_api.fire_simulation(project_gid, session_stored_simulator, self.temp_folder)
 
-    def launch_operation(self, project_gid, algorithm_module, algorithm_classname, view_model):
+    def launch_operation(self, project_gid, algorithm_module, algorithm_classname, view_model, data_file_path_1=None,
+                         data_file_path_2=None):
```
Comment (liadomide) [inline framework_tvb/tvb/interfaces/rest/client/tvb_client.py]: somethins more deneric to allow multiple input files, not just 2

File: framework_tvb/tvb/interfaces/rest/client/examples/launch_operation.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+from tvb.config.init.datatypes_registry import populate_datatypes_registry
+from tvb.core.entities.model.model_operation import STATUS_ERROR, STATUS_CANCELED, STATUS_FINISHED
+from tvb.core.entities.storage import dao
+from tvb.core.neocom import h5
+from tvb.core.neotraits._h5core import ViewModelH5
+from tvb.interfaces.rest.client.tvb_client import TVBClient
+
+if __name__ == '__main__':
+    # TODO: remove this line after pypi package update
+    populate_datatypes_registry()
```
Comment (liadomide) [inline framework_tvb/tvb/interfaces/rest/client/examples/launch_operation.py]: This line (and the import) should be removed, as it is already present in the TVBClient

File: framework_tvb/tvb/interfaces/rest/client/examples/fire_simulation.py [hunk 4]
```diff
@@ -30,16 +30,21 @@
...
 from tvb.adapters.simulator.simulator_adapter import SimulatorAdapterModel
 from tvb.basic.logger.builder import get_logger
+from tvb.config.init.datatypes_registry import populate_datatypes_registry
 from tvb.core.entities.model.model_operation import STATUS_FINISHED, STATUS_CANCELED, STATUS_ERROR
 from tvb.core.neocom import h5
 from tvb.interfaces.rest.client.tvb_client import TVBClient
 
 if __name__ == '__main__':
+    # TODO: Remove this line after pypi package update
+    populate_datatypes_registry()
```
Comment (liadomide) [inline framework_tvb/tvb/interfaces/rest/client/examples/fire_simulation.py]: Fix as written in the TODO

File: framework_tvb/tvb/interfaces/rest/server/resources/operation/operation_resource.py [hunk 5]
```diff
@@ -117,10 +121,24 @@ def post(self, project_gid, algorithm_module, algorithm_classname):
                                                                  {})
             storage_path = self.files_helper.get_project_folder(project, str(operation.id))
 
+            if isinstance(adapter_instance, ABCUploader):
+
+                index = 1
+                for key, value in adapter_instance.get_upload_information().items():
+                    data_file = self.extract_file_from_request(file_name='data_file_' + str(index), file_extension=value)
```
Comment (liadomide) [inline framework_tvb/tvb/interfaces/rest/server/resources/operation/operation_resource.py]: Can;t you already use the 'key' from the Uploader. get_upload_information on the client SDK also?

**Comment by VinczeRobert** [comment]
Then the client will also have to know the 'key' from the Uploader in order
to attach it to the request

**Comment by liadomide** [comment]
Yes, but the client already needs to import the ViewModel for this adapter, so importing something else from the same python modules does not add much complexity.
Suggestion: move the get_upload_information static method on the Form, from the Adapter

File: framework_tvb/tvb/interfaces/rest/client/examples/launch_operation.py [hunk 6]
```diff
@@ -95,12 +91,13 @@
...
     surface_view_model.should_center = False
+    file_name = ZIPSurfaceImporterForm.get_upload_information().keys()
+    data_files = {list(file_name)[0]: surface_zip_path}
 
     logger.info("Launching surface uploading operation...")
     algorithm_module = 'tvb.adapters.uploaders.zip_surface_importer'
     algorithm_class = 'ZIPSurfaceImporter'
-    operation_gid = tvb_client.launch_operation(project_gid, algorithm_module, algorithm_class, surface_view_model,
-                                                data_file_path_1=surface_zip_path)
+    operation_gid = tvb_client.launch_operation(project_gid, algorithm_module, algorithm_class, surface_view_model, data_files=data_files)
```
Comment (liadomide) [inline framework_tvb/tvb/interfaces/rest/client/examples/launch_operation.py]: It is ok now, in terms of how generic the code was written, but the API is hard to use, with some redundant information requested from the end-user.
I would like this to be the end call:
`operation_gid = tvb_client.launch_operation(project_gid, ZIPConnectivityImporter, connectivity_view_model)`
And we should deal with all the strings and files selection ourselves

**Pull request merged**